js/plugins/mcp/src/client/client.ts

```diff
@@ -0,0 +1,291 @@
+
+    const { transport, type: transportType } = await transportFrom(config);
+    if (!transport) {
+      throw new GenkitError({
+        status: 'INVALID_ARGUMENT',
+        message: `[MCP Client] Could not determine valid transport config from supplied options.`,
+      });
+    }
+
+    let error: string | undefined;
+
+    const client = new Client({ name: this.name, version: this.version });
+    client.registerCapabilities({ roots: {} });
+
+    if (this.isEnabled()) {
+      try {
+        await client.connect(transport);
+      } catch (e) {
+        logger.warn(
+          `[MCP Client] Error connecting server via ${transportType} transport: ${e}`
+        );
+        this.disabled = true;
+        error = (e as Error).toString();
+      }
+    }
+
+    this._server = {
+      client,
+      transport,
+      error,
+    } as McpServerRef;
+  }
+
+  /**
+   * Disconnects the MCP server and removes its registration
+   * from this client instance.
+   */
+  async disconnect() {
+    if (this._server) {
+      logger.info(
+        `[MCP Client] Disconnecting MCP server in client '${this.name}'.`
+      );
+      await this._server.client.close();
+      this._server = undefined;
+    }
+  }
+
+  /**
+   * Temporarily disables a server connection. Closes the underlying transport
+   * but retains the server's configuration. Does nothing if the server is
+   * already disabled.
+   */
+  async disable() {
+    if (!this.isEnabled()) return;
+    if (this._server) {
+      logger.info(`[MCP Client] Disabling MCP server in client '${this.name}'`);
+      this.disabled = true;
+    }
+  }
+
+  /**
+   * Whether this client-server connection is enabled or not.
+   */
+  isEnabled() {
+    return !this.disabled;
+  }
+
+  /**
+   * Re-enables a previously disabled server connection. Does nothing if the
+   * server is not disabled.
+   */
+  async reenable() {
+    if (this.isEnabled()) return;
+    if (this._server) {
+      logger.info(
+        `[MCP Client] Reenabling MCP server in client '${this.name}'`
+      );
+      this.disabled = false;
+    }
+  }
+
+  /**
+   * Closes and then restarts the transport connection for the specified server.
+   * Useful for attempting to recover from connection issues without full
+   * reconfiguration.
+   */
+  async reconnect() {
+    if (this._server) {
+      logger.info(
+        `[MCP Client] Reconnecting MCP server in client '${this.name}'`
+      );
+      await this.disconnect();
+      await this.connect(this.serverConfig);
+    }
+  }
+
+  async getTools(ai: Genkit): Promise<ToolAction[]> {
+    await this.ready();
+    let tools: ToolAction[] = [];
+
+    if (this._server) {
+      const capabilities = await this._server.client.getServerCapabilities();
+      if (capabilities?.tools)
+        tools.push(
+          ...(await fetchDynamicTools(ai, this._server.client, {
+            rawToolResponses: this.rawToolResponses,
+            serverName: this.name + '-mcp-server',
```

Comment:
I don't think we need to append `-mcp-server` here do we?

Reply:
I don't know what I was thinking. Removed.